lib/privileges.ts

```diff
@@ -0,0 +1,272 @@
+ * @param privileges Array of privileges to be connect to user
+ * @returns
+ */
+export const updatePrivilegesForUser = async (
+  ability: Ability,
+  userID: string,
+  privileges: { id: string; action: "add" | "remove" }[]
+) => {
+  try {
+    checkPrivileges(ability, [{ action: "update", subject: "User" }]);
+    const addPrivileges: { id: string }[] = [];
+    const removePrivileges: { id: string }[] = [];
+    privileges.forEach((privilege) => {
+      if (privilege.action === "add") {
+        addPrivileges.push({ id: privilege.id });
+      } else {
+        removePrivileges.push({ id: privilege.id });
+      }
+    });
+    const user = await prisma.user.update({
+      where: {
+        id: userID,
+      },
+      data: {
+        privileges: {
+          connect: addPrivileges,
+          disconnect: removePrivileges,
+        },
+      },
+      select: {
+        privileges: true,
+      },
+    });
+    await privilegeDelete(userID);
+
+    return user.privileges;
+  } catch (error) {
+    logMessage.error(error as Error);
+    if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === "P2025") {
+      // Error P2025: Record to update not found.
+      return null;
+    }
+    throw error;
+  }
+};
+
+/**
+ * Get all privileges availabe in the application
+ * @returns an array of privealges
+ */
+export const getAllPrivileges = async (ability: Ability) => {
+  try {
+    checkPrivileges(ability, [{ action: "view", subject: "Privilege" }]);
+    return await prisma.privilege.findMany({
+      select: {
+        id: true,
+        nameEn: true,
+        nameFr: true,
+        descriptionEn: true,
+        descriptionFr: true,
+        permissions: true,
+        priority: true,
+      },
+      orderBy: {
+        priority: "asc",
+      },
+    });
+  } catch (e) {
+    return prismaErrors(e, []);
+  }
+};
+
+export const updatePrivilege = async (ability: Ability, privilege: Privilege) => {
+  try {
+    checkPrivileges(ability, [{ action: "update", subject: "Privilege" }]);
+
+    const response = await prisma.privilege.update({
+      where: {
+        id: privilege.id,
+      },
+      data: privilege,
+
+      select: {
+        id: true,
+      },
+    });
+    // Flush existing privilege cache for all users asynchronously
+    flushValues();
+    return response;
+  } catch (error) {
+    logMessage.error(error as Error);
+    if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === "P2025") {
+      // Error P2025: Record to update not found.
+      return null;
+    }
+    throw error;
+  }
+};
+
+export const createPrivilege = async (ability: Ability, privilege: Privilege) => {
+  try {
+    checkPrivileges(ability, [{ action: "create", subject: "Privilege" }]);
+
+    const response = await prisma.privilege.create({
+      data: privilege,
+
+      select: {
+        id: true,
+      },
+    });
+    // Flush existing privilege cache for all users asynchronously
+    flushValues();
```

Comment:
No, because the `flushValues()` can be a long running event.  It is not critical that the function resolves before returning.  If the fn rejects the expiry keys on entries will handle the flush over a period up to 10 min.  After 10 min we're guarenteed that the cache is completely refreshed.